akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala

```diff
@@ -113,6 +114,32 @@ abstract class ExtensibleBehavior[T] extends Behavior[T](BehaviorTags.Extensible
   def receiveSignal(ctx: TypedActorContext[T], msg: Signal): Behavior[T]
 }
 
+/**
+ * INTERNAL API
+ * A behavior type that could be supervised, Not for user extension.
+ */
+@InternalApi
+class SuperviseBehavior[T] private[akka] (val wrapped: Behavior[T])
+    extends Behavior[T](BehaviorTags.SuperviseBehavior) {
+  private final val ThrowableClassTag = ClassTag(classOf[Throwable])
+
+  /** Specify the [[SupervisorStrategy]] to be invoked when the wrapped behavior throws. */
+  def whenFailure[Thr <: Throwable](strategy: SupervisorStrategy)(implicit tag: ClassTag[Thr]): SuperviseBehavior[T] = {
+    val effectiveTag = if (tag == ClassTag.Nothing) ThrowableClassTag else tag
+    new SuperviseBehavior[T](Supervisor(Behavior.validateAsInitial(wrapped), strategy)(effectiveTag))
+  }
+
+  /**
+   * Specify the [[SupervisorStrategy]] to be invoked when the wrapped behavior throws.
+   *
+   * Only exceptions of the given type (and their subclasses) will be handled by this supervision behavior.
+   */
+  def whenFailure[Thr <: Throwable](clazz: Class[Thr], strategy: SupervisorStrategy): SuperviseBehavior[T] =
+    whenFailure(strategy)(ClassTag(clazz))
+
+  private[akka] def unwrap: Behavior[T] = wrapped
+}
```

Comment:
Unfortunate that the entire class needs to be public, but maybe separating in public trait and impl would be tricky. Make it final and replace "not for user extension" with "Not for user instantiation.

`@InternalApi` would mean the entire type is internal but we return it so drop that but put it on the unwrap method which becomes public from Java.